### PR TITLE
Swift 3 conversion

### PR DESCRIPTION
**Pull request by gmwood**
https://buzzfeed.atlassian.net/browse/IOS-3541

so this had some sketchy stuff that I can't test (it builds). i deleted the test target because there was a warning on the target and it had no tests, and also clicked update to recommended settings.  I also added a .gitignore b/c it was annoying me. I'll tag the stuff that I had to change that I'm a bit scared of.

### CODE REVIEW

File: ImagePalette/ColorCutQuantizer.swift [hunk 1]
```diff
@@ -228,7 +228,10 @@ private class Vbox: Hashable {
 	}
 
 	var volume: Int {
-		return (self.maxRed - self.minRed + 1) * (self.maxGreen - self.minGreen + 1) * (self.maxBlue - self.minBlue + 1)
+        let red: Int = self.maxRed - self.minRed + 1
+        let green: Int = self.maxGreen - self.minGreen + 1
+        let blue: Int = self.maxBlue - self.minBlue + 1
```
Comment (gmwood) [inline ImagePalette/ColorCutQuantizer.swift:233]: doh i'll fix indentation

File: ImagePalette/Palette.swift [hunk 2]
```diff
@@ -9,9 +9,11 @@
 import Foundation
 import UIKit
 
-private func createQueue() -> dispatch_queue_t {
-	let queue = dispatch_queue_create("palette.generator", DISPATCH_QUEUE_CONCURRENT)
-	dispatch_set_target_queue(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), queue)
+private func createQueue() -> DispatchQueue {
+	let queue = DispatchQueue(label: "palette.generator", attributes: DispatchQueue.Attributes.concurrent)
```
Comment (gmwood) [inline ImagePalette/Palette.swift]: same here

Comment (gmwood) [reply ImagePalette/Palette.swift]: actually went to a meeting in the middle of fixing this , shit will fix it....

File: ImagePalette/SwiftPriorityQueue.swift [hunk 3]
```diff
@@ -112,27 +112,37 @@ public struct PriorityQueue<T: Comparable> {
...
+    public func formIndex(after i: inout Int) {
+        if i + 1 < heap.endIndex {
+            i = i + 1
+        }
+    }
+
+    public func index(after i: Int) -> Int {
+        return i + 1
+    }
+
```
Comment (gmwood) [inline ImagePalette/SwiftPriorityQueue.swift:145]: the above 2 things I had to add to get rid of an error (conform to IndexableBase). seems right?

File: ImagePalette/UIImage+PaletteExtension.swift [hunk 4]
```diff
@@ -50,26 +50,25 @@ extension UIImage {
...
-                    let pixelInfo = ((pixelsWide * y) + x) * 4
-
-                    let alpha = Int64(data[pixelInfo])
-                    let red = Int64(data[pixelInfo + 1])
-                    let green = Int64(data[pixelInfo + 2])
-                    let blue = Int64(data[pixelInfo + 3])
-
-                    pixels.append(HexColor.fromARGB(alpha, red: red, green: green, blue: blue))
+            let unconstrainedData = context.data
+            if let data = UnsafePointer<UInt8>(unconstrainedData?.assumingMemoryBound(to: UInt8.IntegerLiteralType.self)) {
```
Comment (gmwood) [inline ImagePalette/UIImage+PaletteExtension.swift:58]: this was the part I'm most unsure about.

File: ImagePalette/ColorCutQuantizer.swift (hunk 1 above)
Comment (gmwood) [reply ImagePalette/ColorCutQuantizer.swift:233]: actually the line above is a tab, not on me.

File: ImagePalette/HSLColor.swift [hunk 5]
```diff
@@ -34,7 +34,7 @@ internal class HSLColor {
 		} else {
 			let c = (1.0 - abs(2 * self.lightness - 1.0)) * self.saturation
 			let m = self.lightness - 0.5 * c
-			let x = c * (1.0 - abs((self.hue / 60.0 % 2.0) - 1.0))
+			let x = c * (1.0 - abs(((self.hue / 60.0).truncatingRemainder(dividingBy: 2.0)) - 1.0))
```
Comment (scelis) [inline ImagePalette/HSLColor.swift:37]: 😳

File: ImagePalette.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -154,41 +120,19 @@
...
-		};
 /* End PBXNativeTarget section */
 
 /* Begin PBXProject section */
 		DB02E6E01CA4320D00803AED /* Project object */ = {
 			isa = PBXProject;
 			attributes = {
 				LastSwiftUpdateCheck = 0730;
-				LastUpgradeCheck = 0730;
+				LastUpgradeCheck = 0820;
```
Comment (scelis) [inline ImagePalette.xcodeproj/project.pbxproj:132]: Should we be using Xcode 8.1 as we discussed in slack?

File: ImagePalette.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -373,14 +299,15 @@
 				PRODUCT_NAME = "$(TARGET_NAME)";
 				SKIP_INSTALL = YES;
 				SWIFT_OPTIMIZATION_LEVEL = "-Onone";
-				SWIFT_VERSION = 2.3;
+				SWIFT_VERSION = 3.0;
```
Comment (scelis) [inline ImagePalette.xcodeproj/project.pbxproj]: 3.0.1?

File: ImagePalette/SwiftPriorityQueue.swift (hunk 3 above)
Comment (scelis) [reply ImagePalette/SwiftPriorityQueue.swift:145]: Yeah, I dunno. Why does one of these methods test endIndex but the other does not?

File: ImagePalette/UIImage+PaletteExtension.swift (hunk 4 above)
Comment (scelis) [reply ImagePalette/UIImage+PaletteExtension.swift:58]: Yeah, this seems weird to me. Have you tried running this code to make sure it does what it used it?

File: ImagePalette/SwiftPriorityQueue.swift (hunk 3 above)
Comment (gmwood) [reply ImagePalette/SwiftPriorityQueue.swift:145]: Index has:

`Precondition: (startIndex..<endIndex).contains(i)`

File: ImagePalette/UIImage+PaletteExtension.swift (hunk 4 above)
Comment (gmwood) [reply ImagePalette/UIImage+PaletteExtension.swift:58]: That is the problem, I'd need Quizzes and its dependencies on Swift 3 right? That is the only thing this library is used for, to get a color of an image there, I think to match it w/ some design elements.

File: ImagePalette/HSLColor.swift (hunk 5 above)
Comment (gmwood) [reply ImagePalette/HSLColor.swift:37]: I didn't do anything here, it probably works.

File: ImagePalette.xcodeproj/project.pbxproj (hunk 6 above)
Comment (gmwood) [reply ImagePalette.xcodeproj/project.pbxproj:132]: oh yeah, i can change that stuff

**Comment by gmwood** [comment]
should be xcode 8.1/Swift 3.0.1 now.... maybe I should just leave this open until Quizzes is converted?

File: ImagePalette/UIImage+PaletteExtension.swift (hunk 4 above)
Comment (scelis) [reply ImagePalette/UIImage+PaletteExtension.swift:58]: You could try running the code in an example project or playground if that's easier than waiting for Quizzes

Comment (gmwood) [reply ImagePalette/UIImage+PaletteExtension.swift:58]: I tried in a playground and I got the same results as if I used the same image in Swift 2.3 in the BF app. That cool? I can add the playground to the project if we want.

Comment (scelis) [reply ImagePalette/UIImage+PaletteExtension.swift:58]: sounds good to me. a playground in here would be useful

**Review by scelis — APPROVED** [review]

**Review by ChadEBrady — APPROVED** [review]

**Pull request merged**